File: venti/src/units.rs
//! Unit converters between SI (venti native) and US customary units.

use crate::Result;
pub const CFM_TO_M3S: f64 = 0.0004719474432; // ft^3/min -> m^3/s
pub const INWC_TO_PA: f64 = 249.0889; // inch H2O (4 C) -> Pa
pub const FT_TO_M: f64 = 0.3048;
pub const IN_TO_M: f64 = 0.0254;
pub const FPM_TO_MS: f64 = 0.00508;

pub fn cfm_to_m3s(cfm: f64) -> f64 {
    cfm * CFM_TO_M3S
}
pub fn m3s_to_cfm(m3s: f64) -> f64 {
    m3s / CFM_TO_M3S
}
pub fn inwc_to_pa(inwc: f64) -> f64 {
    inwc * INWC_TO_PA
}
pub fn pa_to_inwc(pa: f64) -> f64 {
    pa / INWC_TO_PA
}
pub fn ft_to_m(ft: f64) -> f64 {
    ft * FT_TO_M
}
pub fn m_to_ft(m: f64) -> f64 {
    m / FT_TO_M
}
pub fn in_to_m(inches: f64) -> f64 {
    inches * IN_TO_M
}
pub fn m_to_in(m: f64) -> f64 {
    m / IN_TO_M
}
pub fn fpm_to_ms(fpm: f64) -> f64 {
    fpm * FPM_TO_MS
}
pub fn ms_to_fpm(ms: f64) -> f64 {
    ms / FPM_TO_MS
}
pub fn f_to_c(fahrenheit: f64) -> f64 {
    (fahrenheit - 32.0) * 5.0 / 9.0
}
pub fn c_to_f(celsius: f64) -> f64 {
    celsius * 9.0 / 5.0 + 32.0
}

/// ACH = (flowrate × 3600) / room_volume.
pub fn air_changes_per_hour(flowrate_m3s: f64, volume_m3: f64) -> Result<f64> {
    if volume_m3 <= 0.0 {
        return Err("volume_m3 must be positive".into());
    }
    if flowrate_m3s < 0.0 {
        return Err("flowrate_m3s must be non-negative".into());
    }
    Ok(flowrate_m3s * 3600.0 / volume_m3)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cfm_roundtrip() {
        let x = 500.0;
        assert!((m3s_to_cfm(cfm_to_m3s(x)) - x).abs() < 1e-9);
    }

    #[test]
    fn inwc_pa_value() {
        // 1 inWC ~ 249 Pa.
        assert!((inwc_to_pa(1.0) - 249.0889).abs() < 1e-9);
    }

    #[test]
    fn temperature_conversions() {
        assert!((c_to_f(100.0) - 212.0).abs() < 1e-9);
        assert!((f_to_c(212.0) - 100.0).abs() < 1e-9);
        assert!((c_to_f(0.0) - 32.0).abs() < 1e-9);
    }

    #[test]
    fn ach_value() {
        // 0.1 m^3/s in a 100 m^3 room = 3.6 ACH.
        assert!((air_changes_per_hour(0.1, 100.0).unwrap() - 3.6).abs() < 1e-9);
    }

    #[test]
    fn ach_rejects_bad_volume() {
        assert!(air_changes_per_hour(0.1, 0.0).is_err());
        assert!(air_changes_per_hour(-1.0, 100.0).is_err());
    }
}

File: csharp/Wenta.Core/Units.cs
using System;

namespace Wenta
{
    /// <summary>Unit converters between SI (wenta's native units) and US customary.
    /// Port of `wenta.units` / `wentamojo.units` — constants identical.</summary>
    public static class Units
    {
        private const double CFM_TO_M3S = 0.0004719474432; // ft^3/min -> m^3/s
        private const double INWC_TO_PA = 249.0889;       // inch H2O (4 °C) -> Pa
        private const double FT_TO_M = 0.3048;
        private const double IN_TO_M = 0.0254;
        private const double FPM_TO_MS = 0.00508;

        public static double CfmToM3s(double cfm) { return cfm * CFM_TO_M3S; }
        public static double M3sToCfm(double m3s) { return m3s / CFM_TO_M3S; }
        public static double InwcToPa(double inwc) { return inwc * INWC_TO_PA; }
        public static double PaToInwc(double pa) { return pa / INWC_TO_PA; }
        public static double FtToM(double ft) { return ft * FT_TO_M; }
        public static double MToFt(double m) { return m / FT_TO_M; }
        public static double InToM(double inches) { return inches * IN_TO_M; }
        public static double MToIn(double m) { return m / IN_TO_M; }
        public static double FpmToMs(double fpm) { return fpm * FPM_TO_MS; }
        public static double MsToFpm(double ms) { return ms / FPM_TO_MS; }
        public static double FToC(double fahrenheit) { return (fahrenheit - 32.0) * 5.0 / 9.0; }
        public static double CToF(double celsius) { return celsius * 9.0 / 5.0 + 32.0; }

        /// <summary>ACH = flowrate × 3600 / room volume.</summary>
        public static double AirChangesPerHour(double flowrateM3s, double volumeM3)
        {
            if (volumeM3 <= 0.0)
                throw new WentaException("volume_m3 must be positive");
            if (flowrateM3s < 0.0)
                throw new WentaException("flowrate_m3s must be non-negative");
            return flowrateM3s * 3600.0 / volumeM3;
        }
    }

    /// <summary>Exception mirroring Python's ValueError contract.</summary>
    public class WentaException : Exception
    {
        public WentaException(string message) : base(message) { }
    }
}
